The analysis posted below is a baseline analysis of the Titanic Data Set using simple functions in numpy and pandas.

In [2]:
import numpy as np
import pandas as pd
titanic=pd.read_csv("train.csv")
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [3]:
titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [4]:
len (titanic.index)

891

In [5]:
df2=titanic.drop(columns=['Cabin'])
df2
df2=df2.dropna(subset=['Embarked'])
df2

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
888      2        W./C. 6607  23.4500        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[889 rows x 11 columns]

In [6]:
age_analysis = df2.groupby('SibSp')['Age'].agg(['count', 'mean']).reset_index()
age_analysis

SibSp  count       mean
0      0    469  31.318230
1      1    183  30.089727
2      2     25  22.620000
3      3     12  13.916667
4      4     18   7.055556
5      5      5  10.200000
6      8      0        NaN

In [7]:
df2.loc[df2.Age.isnull(),'Age'] = df2.groupby('SibSp').Age.transform('mean').round(1)
df2

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  30.1      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
888      2        W./C. 6607  23.4500        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[889 rows x 11 columns]

In [8]:
bins = [0,10,20,30,40,50,60,90]
bin_labels = ['<10','10-19','20-29', '30-39', '40-49', '50-59', '>60']
df2['AgeGr'] = pd.cut(df2.Age, bins, right=False, labels = bin_labels)

In [9]:
age_analysis = df2.groupby('AgeGr')['Parch'].agg(['count', 'mean','sum']).reset_index()
age_analysis

AgeGr  count      mean  sum
0    <10     62  1.403226   87
1  10-19    106  0.490566   52
2  20-29    223  0.246637   55
3  30-39    329  0.209726   69
4  40-49     89  0.471910   42
5  50-59     48  0.270833   13
6    >60     25  0.320000    8

I engineered the attribute, AgeGr (Age Group). I did this because the age category was not concise and I felt that the Age was an important attribute to include. So, I created Age Group, which assigns each Age into a category (<10, 10-19, 20-29, 30-39, 40-49, 50-59, >60) in order to provide a snapshot into the survival rate of each age group. I used AgeGr as an alternative to Age.

In [10]:
sexage_Analysis = pd.crosstab(df2['Sex'],df2['AgeGr'],df2['Survived'],aggfunc='mean').round(1)
sexage_Analysis

AgeGr   <10  10-19  20-29  30-39  40-49  50-59  >60
Sex                                                
female  0.6    0.7    0.7    0.8    0.7    0.9  1.0
male    0.6    0.1    0.2    0.2    0.2    0.1  0.1

In [11]:
sexclass_Analysis = pd.crosstab(df2['Sex'],df2['Pclass'],df2['Survived'],aggfunc='mean')
sexclass_Analysis

Pclass         1         2         3
Sex                                 
female  0.967391  0.921053  0.500000
male    0.368852  0.157407  0.135447

In [12]:
ageclass_Analysis = pd.crosstab(df2['AgeGr'],df2['Pclass'],df2['Survived'],aggfunc='mean')
ageclass_Analysis

Pclass         1         2         3
AgeGr                               
<10     0.666667  1.000000  0.452381
10-19   0.833333  0.500000  0.242857
20-29   0.705882  0.415094  0.242647
30-39   0.658228  0.389831  0.246073
40-49   0.594595  0.555556  0.058824
50-59   0.555556  0.333333  0.000000
>60     0.250000  0.250000  0.200000

In [13]:
ageparch_Analysis = pd.crosstab(df2['AgeGr'],df2['Parch'],df2['Survived'],aggfunc='mean')
ageparch_Analysis

Parch         0         1         2    3    4     5    6
AgeGr                                                   
<10    1.000000  0.657143  0.538462  NaN  NaN   NaN  NaN
10-19  0.380282  0.315789  0.533333  0.0  NaN   NaN  NaN
20-29  0.310526  0.533333  0.666667  1.0  0.0   NaN  NaN
30-39  0.346021  0.608696  0.538462  NaN  NaN  0.25  NaN
40-49  0.373134  0.571429  0.333333  0.0  0.0  0.00  0.0
50-59  0.368421  0.625000  0.000000  1.0  NaN   NaN  NaN
>60    0.250000  0.250000       NaN  NaN  0.0   NaN  NaN

In [14]:
Parchclass_Analysis = pd.crosstab(df2['Parch'],df2['Pclass'],df2['Survived'],aggfunc='mean')
Parchclass_Analysis

Pclass         1         2         3
Parch                               
0       0.602484  0.358209  0.225722
1       0.677419  0.750000  0.363636
2       0.761905  0.812500  0.255814
3            NaN  1.000000  0.333333
4       0.000000       NaN  0.000000
5            NaN       NaN  0.200000
6            NaN       NaN  0.000000

In [15]:
Sexparch_Analysis = pd.crosstab(df2['Sex'],df2['Parch'],df2['Survived'],aggfunc='mean')
Sexparch_Analysis

Parch          0         1         2     3    4     5    6
Sex                                                       
female  0.786458  0.766667  0.612245  0.75  0.0  0.25  0.0
male    0.165289  0.327586  0.322581  0.00  0.0  0.00  NaN

According to my analysis, the most likely group to survive are younger females, who have a small family and sat in first class. For my aggregate statistic, I used the mean because the data was binary and with binary data, you get the chance of survival, which is very useful for such an analysis. We can see this through the many dataframes. As for analysis in regards to Sex and Class, where a woman in first class has a 96.7% chance of survival, whereas a male in the same class has a 36.9% chance of survival. This observation is validated by looking at the other classes and realizing that in general, women are more likely to survive and the higher the class, the higher the survival rate would be. As for Age and Sex, women had a very high chance of survival, have a survival rate between 70%-100%, whereas men older than 10 years, had a low chance of survival, having a survival rate between 10%-20%. The Age and Class Analysis further validate the findings stated that the higher the class, the higher the chance at survival. As for families, I used the parent/child statistic because it had more representation throughout the earlier stated categories. In this, I found that the smaller the family, the higher the chance at survival. This can be seen in all tables above with pertinence to Parch.

I excluded PassengerID, Name, Ticket, Fare, Cabin and Embarked. I excluded PassengerID because there was no real correlation in this category, so I felt that the information gathered from this category would not provide pertinent information. I excluded Name and Ticket for a similar reason as PassengerID, as it did not provide pertinent information for the analysis needed. Fare was not included because it provided similar insight as Pclass (the more one paid for a ticket meant the higher the class they sat) but the values were a bit less concise. I did not include Cabin because this category had around 77% of the data missing, which meant that the analysis of this data would not lead to a good conclusion about the Titanic in general. Finally, I left out Embarked because it did not appear to heavily influence the survival rate (ie: where you embarked did not have as much of a say in whether or not someone survived in my opinion).

As for the missing values in the data, I treated each category different, as each category represented a different scenario. The three categories with missing data were Cabin, Embarked and Age. For Cabin, I realized that 77% of the data was missing. Since there was so much information missing, it would be really difficult for me to predict the missing values, so I felt that it did not have much pertinence for my data analysis, so I deleted the column. For the missing Embarked values, since it was less than 1%, 2/891 passengers, I deleted those entries. For Age, I took a different approach. I realized that around 15% of the data was missing. So, I looked at the other completed attribute to see which one predicted age the best. I realized the SibSp was a good indicator since the mean age for each section of SibSp accurately predicted the Age of a prospective person in such a scenario. For example, if someone were travelling alone or with one Sibling/Spouse, they were probably a person in their 20s or 30s, which was accurately predicted by the SibSp attribute. I did recognize that there would be some degree of error, but since Age was such a vital attribute in my prediction model, I chose to use this prediction in hope that the error would be minimized and that the chance of survival in pertinence to Age would be known.